### PR TITLE
Agent and Account jobs

### PR DESCRIPTION
**Pull request by furkhat**
Partially resolves BV-150. 

Implements agent and account job handlers.
Moved messages (communication with SOMC) related packages to `message` package.
Moved job related packages to `job` package, moved job queue to `job/queue`.

### CODE REVIEW

File: data/misc.go [hunk 1]
```diff
@@ -18,8 +20,51 @@ func FromBytes(src []byte) string {
...
+// Uint32ToBytes using big endian.
+func Uint32ToBytes(x uint32) [4]byte {
+	var xBytes [4]byte
+	binary.BigEndian.PutUint32(xBytes[:], x)
+	return xBytes
+}
+
+// HashPassword computes encoded hash of the password.
+func HashPassword(password, salt string) (string, error) {
+	salted := []byte(fmt.Sprint(password, salt))
```
Comment (dzeckelev) [inline data/misc.go]: maybe we will not use reflect where it is not needed? 
salted := []byte(password + salt)

File: job/handler/agent_test.go [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+	select {
+	case ret := <-somcEndpointChan:
+		if ret.Channel != fixture.Endpoint.Channel {
+			t.Fatal("wrong channel used to publish endpoint")
+		}
+		msgBytes := data.TestToBytes(t, fixture.Endpoint.RawMsg)
+		if !bytes.Equal(msgBytes, ret.Endpoint) {
+			t.Fatal("wrong endpoint sent to somc")
+		}
+	case <-time.After(10 * time.Second):
```
Comment (dzeckelev) [inline job/handler/agent_test.go]: The timeout can be set in the test configuration. You use it in two places

File: job/handler/common_test.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+func TestAfterAccountAddBalance(t *testing.T) {
+	// update balance in DB.accounts.ptc_balance
+	env := newHandlerTest(t)
+	fixture := env.newTestFixture(t, data.JobAfterAccountAddBalance,
+		data.JobAccount)
+	defer env.close()
+	defer fixture.close()
+
+	env.ethBack.balancePTC = big.NewInt(100)
```
Comment (dzeckelev) [inline job/handler/common_test.go]: It can be put to the test configuration?

Comment (furkhat) [reply job/handler/common_test.go]: nope, that state is specific for this particular test.

File: uisrv/offering.go [hunk 4]
```diff
@@ -96,24 +96,17 @@ func (s *Server) fillOffering(offering *data.Offering) error {
...
-	err := s.db.FindByPrimaryKeyTo(agent, offering.Agent)
-	if err != nil {
+	if err := s.db.FindByPrimaryKeyTo(agent, offering.Agent); err != nil {
 		return err
 	}
+
+	offering.OfferStatus = data.OfferRegister
+	offering.Status = data.MsgUnpublished
 	offering.Agent = agent.EthAddr
-	// TODO: fix this
```
Comment (sofabeat) [inline uisrv/offering.go]: Leave this comment. We should fix it in future.

File: dappctrl.config.json [hunk 5]
```diff
@@ -41,10 +41,7 @@
         "Level": "info"
     },
 
-    "AgentServer": {
-        "Addr": "localhost:3000",
-        "TLS": null
-    },
+    "PayAddress": "http://localhost:9000/v1/pmtChannel/pay",
```
[the remark this answers is not in the document]
Comment (ababo) [reply dappctrl.config.json]: If its URL then better to use URL-suffix.

File: data/test.go [hunk 6]
```diff
@@ -263,37 +349,76 @@ func CleanTestDB(t *testing.T, db *reform.DB) {
...
+		Channel:        ch,
+		Endpoint:       endp,
 	}
 }
 
 func (f *TestFixture) Close() {
-	DeleteFromTestDB(f.T, f.DB, f.Channel,
-		f.Offering, f.Template, f.Account, f.Product)
+	// (t, db, endpTmpl, prod, acc, user, tmpl, off, ch, endp)
+	DeleteFromTestDB(f.T, f.DB, f.Endpoint, f.Channel, f.Offering,
```
Comment (ababo) [inline data/test.go]: Please delete all the objects in the exactly reversed order of creation.

File: job/handler/adapter_test.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+	offeringHash [32]byte, minDeposit *big.Int, maxSupply uint16) error {
+	b.callStack = append(b.callStack, testEthBackCall{
+		method: "RegisterServiceOffering",
+		caller: opts.From,
+		args:   []interface{}{offeringHash, minDeposit, maxSupply},
+	})
+	return nil
+}
+
+func (b *testEthBackend) PTCBalanceOf(opts *bind.CallOpts, addr common.Address) (*big.Int, error) {
```
Comment (ababo) [inline job/handler/adapter_test.go]: Can you keep lines under 80 chars?

File: job/handler/agent.go [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+	}
+
+	if err := tx.Commit(); err != nil {
+		tx.Rollback()
+		return fmt.Errorf("unable to commit changes: %v", err)
+	}
+
+	return w.addJob(data.JobAgentPreEndpointMsgCreate,
+		data.JobChannel, channel.ID)
+}
```
Comment (ababo) [inline job/handler/agent.go]: Consider splitting this function, it's too big.

File: dappctrl.config.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (furkhat) [reply dappctrl.config.json]: This path send to client with other endpoint information.
I decided to keep it simple and put as it is.
( I don't know where agent wants his payment server. May be the agent will have some reverse proxy on the way to pay server? I don't know. )

File: data/schema.go [hunk 9]
```diff
@@ -112,11 +112,11 @@ type Offering struct {
 	OfferStatus        string  `json:"offerStatus" reform:"offer_status"`
 	BlockNumberUpdated uint64  `json:"blockNumberUpdated" reform:"block_number_updated"`
 	Agent              string  `json:"agent" reform:"agent" validate:"required"`
-	Signature          string  `json:"signature" reform:"signature"` // Agent's signature.
+	RawMsg             string  `json:"rawMsg" reform:"raw_msg"`
```
[the remark this answers is not in the document]
Comment (furkhat) [reply data/schema.go]: [`AgentPreOfferingMsgSOMCPublish`](
https://github.com/Privatix/dappctrl/blob/aad1a933aaa719ae37bd6478a79072689493532c/job/handler/agent.go#L541)

**Pull request merged**